lib/utils/file_transfer.rb

```diff
@@ -0,0 +1,76 @@
+require "services"
+require "utils/file_transfer_error"
+
+module Utils
+  # Use rclone system calls to transfer files between
+  # local and remote storage (and vice versa).
+  class FileTransfer
+    def initialize
+      # Check that config is set...
+      if Settings.rclone_config_path.nil?
+        raise "Settings.rclone_config_path missing from Settings"
+      end
+      # ... and points to an actual file.
+      unless File.exist?(Settings.rclone_config_path)
+        raise "Settings.rclone_config_path points to " \
+              "#{Settings.rclone_config_path}, which does not exist"
+      end
+    end
+
+    # Really just an alias for transfer
+    def upload(local_file, remote_dir)
+      transfer(local_file, remote_dir)
+    end
+
+    # Really just an alias for transfer
+    def download(remote_file, local_dir)
+      transfer(remote_file, local_dir)
+    end
+
+    def exists?(dir)
+      make_call("#{call_prefix} ls \"#{dir}\"")
+    end
+
+    def mkdir(dir)
+      make_call("#{call_prefix} mkdir \"#{dir}\"")
+    end
+
+    # Return parsed JSON ls output
+    def ls_remote_dir(remote_dir)
+      # Never parse ls output, let rclone do it for us.
+      call = "#{call_prefix} lsjson \"#{remote_dir}\""
+      puts "call #{call}"
+      response = `#{call}`
+      puts "response #{response}"
+      JSON.parse(response)
+    rescue JSON::ParserError => _err
+      raise Utils::FileTransferError, "could not ls #{remote_dir}"
```

Comment:
The custom exception class loses the context of the JSON error. I think I might pass `_err` on to the `Utils::FileTransferError` initializer as a custom field and just tuck it away in an ivar for the unlikely situation there's a need to debug unexpected output from `lsjson`. Or more simply, append `_err` in some form a la `"could not ls #{remote_dir} (#{_err})"` (should remove the underscore on `_err` of course yadda yadda)

Reply:
Yeah that's better.